modules/engage-paella-player-7/src/plugins/org.opencast.paella.downloadsPlugin.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ *
+ * The Apereo Foundation licenses this file to you under the Educational
+ * Community License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License
+ * at:
+ *
+ *   http://opensource.org/licenses/ecl2.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+import {
+  createElementWithHtmlText,
+  PopUpButtonPlugin,
+  translate
+} from 'paella-core';
+
+import '../css/DownloadsPlugin.css';
+import defaultDownloadIcon from '../icons/download.svg';
+
+export default class DownloadsPlugin extends PopUpButtonPlugin {
+
+  async isEnabled() {
+    this._downloads = {};
+    const { streams, metadata } = this.player.videoManifest;
+
+    const hasWritePermission = this.config.allowWritePermission && await this.player.opencastAuth.canWrite();
+    const hasAllowedLicense = this.config.allowOnlyLicenses
+      ? (this.config.allowOnlyLicenses.includes(metadata.license))
+      : true;
+    const enabled = (await super.isEnabled()) && (hasWritePermission || hasAllowedLicense);
+
+    if (enabled) {
+      streams.forEach(s => {
+        let streamDownloads = [];
+        const { mp4 } = s.sources;
+        if (mp4) {
```

Comment:
Have you tried what happens if paella is playing a HLS stream? From what I read in the code only works for mp4 files as the source.

Maybe could it be better to left this as a configuration, to make it future-proof.

Reply:
Yes, only mp4 is supported. For HLS, I can put the link to the m3u8 file, but you will be missing all the .ts files.

I think it is not useful for you to download only the m3u8 file.

Reply:
> Understandable, to download a HLS you have to collect all the `ts` files first and that is out of scope. What about `webm` or `mkv`?

On second thought, What do you think if we show for download any file that includes a certain configurable tag?
Only media files? or also attachments?

Reply:
@mliradelc now `webm` or `mkv` are supported. I added a documentation. Please review again.